terratorch/models/smp_model_factory.py

```diff
@@ -1,83 +1,387 @@
 # Copyright contributors to the Terratorch project
 
-"""
-This is just an example of a possible structure to include SMP models
-Right now it always returns a UNET, but could easily be extended to many of the models provided by SMP.
-"""
+import importlib
+from collections.abc import Callable
 
 import segmentation_models_pytorch as smp
+import torch
+import torch.nn.functional as F  # noqa: N812
+from segmentation_models_pytorch.encoders import encoders as smp_encoders
 from torch import nn
 
+from terratorch.datasets import HLSBands
 from terratorch.models.model import Model, ModelFactory, ModelOutput, register_factory
 
 
+class SMPDecoderForPrithviWrapper(nn.Module):
+    """
+    A wrapper for SMP decoders designed to handle single or multiple embeddings.
+
+    Attributes:
+        decoder (nn.Module): The SMP decoder module being wrapped.
+        channels (int): The number of output channels of the decoder.
+
+    Methods:
+        forward_multiple_embeds(*x) -> torch.Tensor:
+            Forward pass for multiple embeddings.
+        forward_single_embed(x) -> torch.Tensor:
+            Forward pass for a single embedding.
+    """
+
+    def __init__(self, decoder, num_channels) -> None:
+        """
+        Args:
+            decoder (nn.Module): The SMP decoder module to be wrapped.
+            num_channels (int): The number of output channels of the decoder.
+        """
+        super().__init__()
+        self.decoder = decoder
+        self.channels = num_channels
+
+    @property
+    def output_embed_dim(self):
+        return self.channels
+
+    def forward_multiple_embeds(self, *x):
+        return self.decoder(*x)
+
+    def forward_single_embed(self, x):
+        return self.decoder(x[-1])
```

Comment:
Instead of this, can we remain consistent with current Prithvi decoders and accept an `in_index` argument? This should be a single int or a list of ints that are the indices of `x` which we select and pass forward to the decoder